contracts/examples/erc1155-marketplace/src/lib.rs

```diff
@@ -0,0 +1,386 @@
+#![no_std]
+#![allow(unused_attributes)]
+
+use elrond_wasm::HexCallDataSerializer;
+
+elrond_wasm::imports!();
+elrond_wasm::derive_imports!();
+
+const ACCEPTED_TRANSFER_ANSWER: u32 = 0xbc197c81;
+
+const TRANSFER_TOKEN_ENDPOINT_NAME: &[u8] = b"safeTransferFrom";
+
+const PERCENTAGE_TOTAL: u8 = 100;
+
+#[derive(TopEncode, TopDecode, TypeAbi)]
+pub struct Auction<BigUint: BigUintApi> {
+	pub min_bid: BigUint,
+	pub max_bid: BigUint,
+	pub deadline: u64,
+	pub original_owner: Address,
+	pub current_bid: BigUint,
+	pub current_winner: Address,
+}
+
+#[derive(TopDecode, TypeAbi)]
+pub struct AuctionArgument<BigUint: BigUintApi> {
+	pub min_bid: BigUint,
+	pub max_bid: BigUint,
+	pub deadline: u64,
+}
+
+#[elrond_wasm_derive::contract(Erc1155MarketplaceImpl)]
+pub trait Erc1155Marketplace {
+	/// `bid_cut_percentage` is the cut that the contract takes from any sucessful bid
+	#[init]
+	fn init(&self, token_ownership_contract_address: Address, bid_cut_percentage: u8) {
+		self.set_token_ownership_contract_address(&token_ownership_contract_address);
+		self.set_percentage_cut(bid_cut_percentage);
+	}
+
+	// endpoints - Token ownership contract only
+
+	/// `_operator` argument is ignored, but it has to be kept because of the erc1155 standard
+	#[endpoint(onERC1155Received)]
+	fn on_erc1155_received(
+		&self,
+		_operator: Address,
+		from: Address,
+		type_id: BigUint,
+		token_id: BigUint,
+		args: AuctionArgument<BigUint>,
+	) -> SCResult<u32> {
+		require!(
+			self.get_caller() == self.get_token_ownership_contract_address(),
+			"Only the token ownership contract may call this function"
+		);
+
+		sc_try!(self.try_create_auction(
+			&type_id,
+			&token_id,
+			&from,
+			&args.min_bid,
+			&args.max_bid,
+			args.deadline
+		));
+
+		Ok(ACCEPTED_TRANSFER_ANSWER)
+	}
+
+	/// Same `AuctionArgument` is used for all tokens  
+	/// `_operator` argument is ignored, but it has to be kept because of the erc1155 standard
+	#[endpoint(onERC1155BatchReceived)]
+	fn on_erc1155_batch_received(
+		&self,
+		_operator: Address,
+		from: Address,
+		type_ids: Vec<BigUint>,
+		token_ids: Vec<BigUint>,
+		args: AuctionArgument<BigUint>,
+	) -> SCResult<u32> {
+		require!(
+			self.get_caller() == self.get_token_ownership_contract_address(),
+			"Only the token ownership contract may call this function"
+		);
+		require!(
+			type_ids.len() == token_ids.len(),
+			"type_ids and token_ids lengths do not match"
+		);
+
+		// Don't have to worry about checking if there are duplicates in the entries,
+		// an error here will revert all storage changes automatically
+		for (type_id, token_id) in type_ids.iter().zip(token_ids.iter()) {
+			sc_try!(self.try_create_auction(
```

Comment:
can we create auction with ESDT as well ? like owner says in what kind of token he wants to receive ?

Reply:
Done